beacon_node/beacon_chain/src/beacon_block_reward.rs

```diff
@@ -0,0 +1,161 @@
+            // Will need to compute for pre-altair block as well
+            0
+        } else {
+            self.compute_beacon_block_proposal_reward_altair(block, pre_state, post_state)?
+        };
+
+        Ok(sync_aggregate_reward
+            .safe_add(proposer_slashing_reward)?
+            .safe_add(attester_slashing_reward)?
+            .safe_add(block_proposal_reward)?)
+    }
+
+    fn compute_beacon_block_sync_aggregate_reward<Payload: ExecPayload<T::EthSpec>>(
+        &self,
+        block: BeaconBlockRef<'_, T::EthSpec, Payload>,
+        state: &BeaconState<T::EthSpec>,
+    ) -> Result<u64, BeaconChainError> {
+        if let Ok(sync_aggregate) = block.body().sync_aggregate() {
+            let (_, proposer_reward_per_bit) = compute_sync_aggregate_rewards(state, &self.spec)
+                .map_err(|_| BeaconChainError::BlockRewardSyncError)?;
+            Ok(sync_aggregate.sync_committee_bits.num_set_bits() as u64 * proposer_reward_per_bit)
+        } else {
+            Ok(0)
+        }
+    }
+
+    fn compute_beacon_block_proposer_slashing_reward<Payload: ExecPayload<T::EthSpec>>(
+        &self,
+        block: BeaconBlockRef<'_, T::EthSpec, Payload>,
+        state: &BeaconState<T::EthSpec>,
+    ) -> Result<u64, BeaconChainError> {
+        let mut proposer_slashing_reward = 0;
+
+        let proposer_slashings = block.body().proposer_slashings();
+
+        for proposer_slashing in proposer_slashings {
+            proposer_slashing_reward.safe_add_assign(
+                state
+                    .get_validator(proposer_slashing.proposer_index() as usize)?
+                    .effective_balance
+                    .safe_div(self.spec.whistleblower_reward_quotient)?,
+            )?;
+        }
+
+        Ok(proposer_slashing_reward)
+    }
+
+    fn compute_beacon_block_attester_slashing_reward<Payload: ExecPayload<T::EthSpec>>(
+        &self,
+        block: BeaconBlockRef<'_, T::EthSpec, Payload>,
+        state: &BeaconState<T::EthSpec>,
+    ) -> Result<u64, BeaconChainError> {
+        let mut attester_slashing_reward = 0;
+
+        let attester_slashings = block.body().attester_slashings();
+
+        for attester_slashing in attester_slashings {
+            for attester_index in get_slashable_indices(state, attester_slashing)? {
+                attester_slashing_reward.safe_add_assign(
+                    state
+                        .get_validator(attester_index as usize)?
+                        .effective_balance
+                        .safe_div(self.spec.whistleblower_reward_quotient)?,
+                )?;
+            }
+        }
+
+        Ok(attester_slashing_reward)
+    }
+
+    fn compute_beacon_block_proposal_reward_altair<Payload: ExecPayload<T::EthSpec>>(
+        &self,
+        block: BeaconBlockRef<'_, T::EthSpec, Payload>,
+        pre_state: &BeaconState<T::EthSpec>,
+        post_state: &BeaconState<T::EthSpec>,
+    ) -> Result<u64, BeaconChainError> {
```

Comment:
I've realised there's an issue with the design I laid out for this API.

The problem with computing a diff between the participation bits is that it doesn't take into account the division-with-rounding that happens _for each attestation_.

E.g. in the spec, `process_attestation` is applied to each attestation individually, and a `proposer_reward_numerator` is computed _for each attestation_.

https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#modified-process_attestation

This can change the end result, because if we have for example 2 attestations, both with `proposer_reward_numerator == 15`, and `proposer_reward_denominator == 10`, then the result of computing the rewards separately for each attestation is:

```python
floor(15/10) + floor(15/10) = 1 + 1 = 2
```

Whereas the result of adding all the `proposer_reward_numerator`s first and then doing the division is different:

```python
floor((15 + 15)/10)) = 3
```
The numbers we're dealing with are much larger, but the same thing can still happen.

To fix this I think we have no choice but to iterate the attestations one by one, something like:

```rust
let mut total_proposer_reward = 0;

let proposer_reward_denominator = WEIGHT_DENOMINATOR
    .safe_sub(PROPOSER_WEIGHT)?
    .safe_mul(WEIGHT_DENOMINATOR)?
    .safe_div(PROPOSER_WEIGHT)?;
for attestation in block.attestations() {
    let data = &attestation.data;
    let inclusion_delay = state.slot().safe_sub(data.slot)?.as_u64();
    let participation_flag_indices =
        get_attestation_participation_flag_indices(state, data, inclusion_delay, spec)?;
    
    let attesting_indices = get_attesting_indices_from_state(state, attestation)?;
    
    let mut proposer_reward_numerator = 0;
    for index in attesting_indices {
        // Check whether the bit was set
        if participation_flag_indices.contains(&flag_index)
            && !validator_participation.has_flag(flag_index)?
        {
            proposer_reward_numerator.safe_add_assign(
                get_base_reward(state, index, base_reward_per_increment, spec)?
                    .safe_mul(weight)?,
            )?;
        }
    }
    
    total_proposer_reward.safe_add_assign(proposer_reward_numerator.safe_div(proposer_reward_denominator))?;
}
```

For more details see: https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/consensus/state_processing/src/per_block_processing/process_operations.rs#L111-L173